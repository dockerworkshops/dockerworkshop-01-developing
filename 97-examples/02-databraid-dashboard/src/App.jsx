import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import ReactGridLayout from 'react-grid-layout';
import { Icon, Sidebar, Segment, Menu } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import './App.css';
import {
  showAddWidgetModal,
  showDashboardSidebar,
  hideDashboardSidebar,
  lockDashboard,
  unlockDashboard,
  saveLayoutChange,
} from './actions';
import ModalAddWidget from './components/ModalAddWidget/';
import WidgetContainer from './components/WidgetContainer/';

const Grid = ReactGridLayout.WidthProvider(ReactGridLayout);

export const AppComponent = (props) => {
  const components = (props.ids).map(widgetId => (
    <div key={widgetId} className="widget-container">
      <WidgetContainer id={widgetId} />
    </div>
  ));

  return (
    <div className="page-container">
      <div
        role="link"
        tabIndex="-1"
        className="side-strip"
        onClick={props.showSidebar ? props.hideDashboardSidebar : props.showDashboardSidebar}
      >
        <Icon name={props.showSidebar ? 'chevron right' : 'ellipsis vertical'} />
      </div>
      <div className="grid-container">
        <Sidebar.Pushable as={Segment}>
          <Sidebar
            as={Menu}
            animation="overlay"
            width="thin"
            direction="right"
            visible={props.showSidebar}
            icon="labeled"
            vertical
            inverted
          >
            <Menu.Item name="add_widget" onClick={props.showAddWidgetModal}>
              <Icon name="add circle" />
              Add Widget
            </Menu.Item>
            {props.ids.length ?
              <Menu.Item
                name="lock-unlock-dashboard"
                onClick={props.locked ? props.unlockDashboard : props.lockDashboard}
              >
                <Icon name={props.locked ? 'unlock' : 'lock'} />
                {props.locked ? 'Unlock' : 'Lock'}
              </Menu.Item>
              : null }
            <Menu.Item name="settings" disabled>
              <Icon name="setting" />
              Settings
            </Menu.Item>
          </Sidebar>
          <Sidebar.Pusher>
            <Segment basic>

              <Grid
                verticalCompact={false}
                className="layout"
                layout={props.layout}
                margin={[2, 2]}
                cols={12}
                rowHeight={30}
                width={1200}
                onLayoutChange={(layout) => { props.saveLayoutChange(layout); }}
              >
                {components}
              </Grid>

            </Segment>
          </Sidebar.Pusher>
        </Sidebar.Pushable>
      </div>
      <ModalAddWidget />
    </div>
  );
};

AppComponent.propTypes = {
  ids: PropTypes.arrayOf(PropTypes.string).isRequired,
  layout: PropTypes.arrayOf(PropTypes.shape({
    i: PropTypes.string.isRequired,
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    w: PropTypes.number.isRequired,
    h: PropTypes.number.isRequired,
    static: PropTypes.bool,
  })).isRequired,
  showSidebar: PropTypes.bool.isRequired,
  locked: PropTypes.bool.isRequired,
  showAddWidgetModal: PropTypes.func.isRequired,
  showDashboardSidebar: PropTypes.func.isRequired,
  hideDashboardSidebar: PropTypes.func.isRequired,
  lockDashboard: PropTypes.func.isRequired,
  unlockDashboard: PropTypes.func.isRequired,
  saveLayoutChange: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => {
  const ids = state.widgets.ids;
  const layout = state.widgets.grid.layout;
  const showSidebar = state.widgets.showSidebar;
  const locked = state.widgets.locked;
  return { ids, layout, showSidebar, locked };
};

export const mapDispatchToProps = dispatch => bindActionCreators({
  showAddWidgetModal,
  showDashboardSidebar,
  hideDashboardSidebar,
  lockDashboard,
  unlockDashboard,
  saveLayoutChange,
},
dispatch);


export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppComponent);
