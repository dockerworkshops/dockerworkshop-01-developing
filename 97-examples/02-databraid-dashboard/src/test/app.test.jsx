import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
/* eslint-disable import/extensions */
import { AppComponent, mapStateToProps, mapDispatchToProps } from '../App.jsx';

const state = {
  widgets: {
    ids: ['github'],
    showSidebar: true,
    grid: {
      layout: [{ i: 'slack', x: 0, y: 8, w: 6, h: 8, static: true }],
    },
  },
};
/* eslint-disable react/jsx-boolean-value */
describe('App component', () => {
  it('should render the App component with props as specified ', () => {
    const showAddWidgetModal = jest.fn();
    const showDashboardSidebar = jest.fn();
    const hideDashboardSidebar = jest.fn();
    const lockDashboard = jest.fn();
    const unlockDashboard = jest.fn();
    const saveLayoutChange = jest.fn();
    const component = shallow(
      <AppComponent
        showAddWidgetModal={showAddWidgetModal}
        showDashboardSidebar={showDashboardSidebar}
        hideDashboardSidebar={hideDashboardSidebar}
        lockDashboard={lockDashboard}
        unlockDashboard={unlockDashboard}
        saveLayoutChange={saveLayoutChange}
        ids={['github']}
        showSidebar={true}
        layout={[{ i: 'slack', x: 0, y: 8, w: 6, h: 8, static: true }]}
        locked={false}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('maps given state to props ', () => {
    const expected = {
      ids: ['github'],
      showSidebar: true,
      layout: [{ i: 'slack', x: 0, y: 8, w: 6, h: 8, static: true }],
    };
    expect(mapStateToProps(state)).toEqual(expected);
  });

  it('maps component dispatches to props ', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('showAddWidgetModal');
    expect(mapDispatchToProps(dispatch)).toHaveProperty('showDashboardSidebar');
    expect(mapDispatchToProps(dispatch)).toHaveProperty('hideDashboardSidebar');
  });

  it('Should have className page-container', () => {
    const showAddWidgetModal = jest.fn();
    const showDashboardSidebar = jest.fn();
    const hideDashboardSidebar = jest.fn();
    const lockDashboard = jest.fn();
    const unlockDashboard = jest.fn();
    const saveLayoutChange = jest.fn();
    const component = shallow(
      <AppComponent
        showAddWidgetModal={showAddWidgetModal}
        showDashboardSidebar={showDashboardSidebar}
        hideDashboardSidebar={hideDashboardSidebar}
        lockDashboard={lockDashboard}
        unlockDashboard={unlockDashboard}
        saveLayoutChange={saveLayoutChange}
        ids={['github']}
        showSidebar={true}
        layout={[{ i: 'slack', x: 0, y: 8, w: 6, h: 8, static: true }]}
        locked={false}
      />,
    );
    expect(component.find('.page-container').exists()).toBe(true);
  });
});
