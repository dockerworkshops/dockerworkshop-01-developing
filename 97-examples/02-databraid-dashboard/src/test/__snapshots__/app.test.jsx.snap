// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render the App component with props as specified  1`] = `
<div
  className="page-container"
>
  <div
    className="side-strip"
    onClick={[Function]}
    role="link"
    tabIndex="-1"
  >
    <Icon
      as="i"
      name="chevron right"
    />
  </div>
  <div
    className="grid-container"
  >
    <SidebarPushable
      as={[Function]}
    >
      <Sidebar
        animation="overlay"
        as={[Function]}
        direction="right"
        icon="labeled"
        inverted={true}
        vertical={true}
        visible={true}
        width="thin"
      >
        <MenuItem
          name="add_widget"
          onClick={[Function]}
        >
          <Icon
            as="i"
            name="add circle"
          />
          Add Widget
        </MenuItem>
        <MenuItem
          name="lock-unlock-dashboard"
          onClick={[Function]}
        >
          <Icon
            as="i"
            name="lock"
          />
          Lock
        </MenuItem>
        <MenuItem
          disabled={true}
          name="settings"
        >
          <Icon
            as="i"
            name="setting"
          />
          Settings
        </MenuItem>
      </Sidebar>
      <SidebarPusher>
        <Segment
          basic={true}
        >
          <_class
            className="layout"
            cols={12}
            layout={
              Array [
                Object {
                  "h": 8,
                  "i": "slack",
                  "static": true,
                  "w": 6,
                  "x": 0,
                  "y": 8,
                },
              ]
            }
            margin={
              Array [
                2,
                2,
              ]
            }
            measureBeforeMount={false}
            onLayoutChange={[Function]}
            rowHeight={30}
            verticalCompact={false}
            width={1200}
          >
            <div
              className="widget-container"
            >
              <Connect(WidgetContainer)
                id="github"
              />
            </div>
          </_class>
        </Segment>
      </SidebarPusher>
    </SidebarPushable>
  </div>
  <Connect(Component) />
</div>
`;
